homeassistant/helpers/entity.py

```diff
@@ -534,6 +540,9 @@ def async_on_remove(self, func: CALLBACK_TYPE) -> None:
     async def _async_registry_updated(self, event):
         """Handle entity registry update."""
         data = event.data
+        if data["action"] == "remove" and data["entity_id"] == self.entity_id:
+            await self.async_removed_from_registry()
+
```

Comment:
We could call `self.async_remove()` here as well to remove the state from the state machine, but that should probably be done in a separate PR.

Reply:
So can you document/explain to me how delete changes cascade throughout the MQTT system? 

These two things can happen:
 - A user indicates that an MQTT device needs to be removed
 - An empty discovery configuration is pushed to a topic that we previously got a configuration from. 

These two cases need to be handled and we should document this in the MQTT integration somewhere how this works.

Reply:
#### Creation / Discovery of MQTT entities / devices by MQTT:
Handled by https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mqtt/discovery.py
When receiving a discovery message, a state is created in the Hass state machine, or updated if it already exists.
(In case of an MQTT device trigger, no state is created).
If `unique_id` is set in the config data, an entity registry entry will be created, or updated if it already exists.
If `device` dict is set in the config data, a device registry will be created, or updated if it already exists.

Discovered MQTT enitities will listen for discovery updates - which must be sent on the same MQTT topic - with help of the MqttDiscoveryUpdated mixin:
https://github.com/home-assistant/home-assistant/blob/e170a11547b716de8cfe46a99473f4dbfcb7f62b/homeassistant/components/mqtt/__init__.py#L1159-L1160

#### Update of MQTT entities / devices by MQTT:
Handled by the MqttDiscoveryUpdated mixin:
https://github.com/home-assistant/home-assistant/blob/e170a11547b716de8cfe46a99473f4dbfcb7f62b/homeassistant/components/mqtt/__init__.py#L1187-L1190

#### Removal of MQTT entities / devices by MQTT:
If an empty discovery message is received, this will result in:
- Removal of the state from the state machine
  https://github.com/home-assistant/home-assistant/blob/e170a11547b716de8cfe46a99473f4dbfcb7f62b/homeassistant/components/mqtt/__init__.py#L1182-L1184
- Entity registry entry will NOT be removed
  This should maybe be changed such that entity registry entries are removed?
- Device registry entry will NOT be removed
  This should maybe be changed such that device registry entries are removed?

#### Removal of MQTT entities / devices by WS:
A WS helper exists to fully remove an MQTT device, which will result in:
- Removal of the states from the state machine (as a result of clearing the discovery topics)
- Entity registry entries related to the device will be removed
- Device registry entry will be removed, including MQTT device triggers
- Any retained MQTT discovery topics for entities are cleared from the broker
  https://github.com/home-assistant/home-assistant/blob/e170a11547b716de8cfe46a99473f4dbfcb7f62b/homeassistant/components/mqtt/__init__.py#L1199-L1204
- Any retained MQTT discovery topics for triggers are cleared from the broker
  https://github.com/home-assistant/home-assistant/blob/e170a11547b716de8cfe46a99473f4dbfcb7f62b/homeassistant/components/mqtt/device_trigger.py#L233-L241